com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl

```diff
@@ -169,164 +169,296 @@ float4 SampleEnv(LightLoopContext lightLoopContext, int index, float3 texCoord,
-}
+#if (defined(COARSE_BINNING) || defined(FINE_BINNING))
 
-void GetCountAndStartTile(PositionInputs posInput, uint lightCategory, out uint start, out uint lightCount)
+// Internal. Do not call directly.
+uint TryFindEntityIndex(inout uint i, uint tile, uint2 zBinRange, uint category, out uint entityIndex)
 {
-    int tileOffset = GetTileOffset(posInput, lightCategory);
+    entityIndex = UINT16_MAX;
 
-#if defined(UNITY_STEREO_INSTANCING_ENABLED)
-    // Eye base offset must match code in lightlistbuild.compute
-    tileOffset += unity_StereoEyeIndex * _NumTileFtplX * _NumTileFtplY * LIGHTCATEGORY_COUNT;
-#endif
+    bool b = false;
+    uint n = TILE_ENTRY_LIMIT;
 
-    // The first entry inside a tile is the number of light for lightCategory (thus the +0)
-    lightCount = g_vLightListGlobal[DWORD_PER_TILE * tileOffset + 0] & 0xffff;
-    start = tileOffset;
-}
+    // This part (1) is loop-invariant. These values do not depend on the value of 'i'.
+    // They will only be computed once per category, not once per function call.
+    const uint tileBufferHeaderIndex = ComputeTileBufferHeaderIndex(tile, category, unity_StereoEyeIndex);
+    const uint tileRangeData         = TILE_BUFFER[tileBufferHeaderIndex]; // {last << 16 | first}
+    const bool isTileEmpty           = tileRangeData == UINT16_MAX;
 
-#ifdef USE_FPTL_LIGHTLIST
+    if (!isTileEmpty) // Avoid wasted work
+    {
+        const uint zBinBufferIndex0 = ComputeZBinBufferIndex(zBinRange[0], category, unity_StereoEyeIndex);
+        const uint zBinBufferIndex1 = ComputeZBinBufferIndex(zBinRange[1], category, unity_StereoEyeIndex);
+        const uint zBinRangeData0   = _zBinBuffer[zBinBufferIndex0]; // {last << 16 | first}
+        const uint zBinRangeData1   = _zBinBuffer[zBinBufferIndex1]; // {last << 16 | first}
+
+        // Recall that entities are sorted by the z-coordinate.
+        // So we can take the smallest index from the first bin and the largest index from the last bin.
+        // TODO: while it appears to work, for a large range, this code can actually fail. :-(
+        // Imagine that we are given the first and the last zBins, and that none of the lights overlap them.
+        // All the lights are in between. So the code below would give us an empty range, which is wrong.
+        // Should we loop over the entire zBinRange? Seems impractical... Maybe change the way we sort?
+        const uint2 tileEntityIndexRange = uint2(tileRangeData  & UINT16_MAX, tileRangeData  >> 16);
+        const uint2 zBinEntityIndexRange = uint2(zBinRangeData0 & UINT16_MAX, zBinRangeData1 >> 16);
+
+        if (IntervalsOverlap(tileEntityIndexRange, zBinEntityIndexRange)) // Avoid wasted work
+        {
+            const uint tileBufferBodyIndex = ComputeTileBufferBodyIndex(tile, category, unity_StereoEyeIndex);
 
-uint GetTileSize()
-{
-    return TILE_SIZE_FPTL;
-}
+            // The part (2) below will be actually executed during every function call.
```

Comment:
Really would like to see the output of the compiler here :).
this part is indeed what is suppose to be executed in the while (try loading) loop. But pretty sure compiler will not be smart enough (and the various if()  may not help) . Anyway we can check when code work.

Reply:
Again, this is a trivial optimization. If it doesn't work then it's a compiler bug.

Reply:
https://en.wikipedia.org/wiki/Loop-invariant_code_motion